apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsColourPickerControl.tsx

```diff
@@ -0,0 +1,145 @@
+import { useEffect, useState } from "react"
+import {
+  Box,
+  Flex,
+  FormControl,
+  HStack,
+  InputGroup,
+  InputLeftAddon,
+  VStack,
+} from "@chakra-ui/react"
+import {
+  ControlProps,
+  RankedTester,
+  rankWith,
+  schemaMatches,
+} from "@jsonforms/core"
+import { useJsonForms, withJsonFormsControlProps } from "@jsonforms/react"
+import { FormLabel, Input } from "@opengovsg/design-system-react"
+import get from "lodash/get"
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+import {
+  convertHexToRgb,
+  generateTheme,
+  normalizeHex,
+} from "~/features/settings/utils"
+import { useColorPalette } from "~/hooks/useColorPalete"
+
+export const jsonFormsColourPickerControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.ColourPickerControl,
+  schemaMatches((schema) => schema.format === "color-picker"),
+)
+
+const THEME_PATHS = [
+  "colors.brand.canvas.default",
+  "colors.brand.canvas.alt",
+  "colors.brand.canvas.backdrop",
+  "colors.brand.canvas.inverse",
+  "colors.brand.interaction.default",
+  "colors.brand.interaction.hover",
+  "colors.brand.interaction.pressed",
+] as const
+
+const JsonFormsColourPickerControl = ({
+  data,
+  label,
+  handleChange,
+  path,
+  description,
+  required,
+  errors,
+}: ControlProps) => {
+  // NOTE: Tint is the colour brightened by 10% successively,
+  // shades are the colour darkened by 10% successively
+  const { tints, colour, shades } = useColorPalette(...convertHexToRgb(data))
+  const ctx = useJsonForms()
+  const [displayedColour, setDisplayedColour] = useState(data)
+
+  // NOTE: whenever data changes, write the corresponding correct tints and shades
+  useEffect(() => {
+    const otherPaths = THEME_PATHS.filter((p) => p !== path)
+    const palette = generateTheme({ tints, colour, shades })
+    otherPaths.map((p) => {
+      handleChange(p, palette[p] ?? "#FFFFFF")
+    })
+  }, [data])
+
+  return (
+    <Box>
+      <VStack gap="1.5rem">
+        <FormControl isRequired={required} isInvalid={!!errors}>
+          <FormLabel description={description} mb={0}>
+            {label}
+          </FormLabel>
+          <HStack spacing="0.75rem" alignItems="center" mt="0.75rem">
+            <InputGroup size="xs" w="fit-content">
+              <InputLeftAddon
+                borderLeftRadius="4px"
+                border="1px solid"
+                borderColor="base.divider.strong"
+                bgColor="interaction.support.disabled"
+                textColor="interaction.support.disabled-content"
+              >
+                #
+              </InputLeftAddon>
+              <Input
+                value={displayedColour.replace("#", "")}
+                placeholder="FFFFFF"
+                w="86px"
+                onChange={(e) => {
+                  const rawString = e.target.value
+                  const parsedHex = rawString.replace(/[^0-9A-Fa-f]/g, "")
```

Comment:
instead of handrolling regex, should we instead either use

1. isHexadecimal from validator library
```ts
import { isHexadecimal } from "validator"
const parsedHex = rawString.split("").filter((c) => isHexadecimal(c)).join("")
```

2. pure JS without regex, for better readability
```ts
const parsedHex =
  rawString
    .toUpperCase()
    .split("")
    .filter((c) => "0123456789ABCDEF".includes(c))
    .join("")
```

Reply:
oh cool, didn't realise we had `validator` installed! TIL!